### ISSUE TITLE
Sampling for large datasets

### ISSUE THREAD

**Issue by gradientsky** [awslabs/autogluon#1672]
Add data sampling to handle extremely large datasets.

**Comment by la-cruche** [comment]
interesting approaches in this repo https://github.com/decile-team/cords

**Comment by willsmithorg** [comment]
Would this apply to tabular only, or all prediction types?

**Comment by Innixma** [comment]
@willsmithorg It could apply to all prediction types, although image for example often does not have such a memory issue since it trains in batches while loading the images into memory in batches rather than all at once.

With Tabular, most models don't have easy ways of doing batch training, and even if they do, it might be much slower than training in-memory.

**Comment by willsmithorg** [comment]
A few questions / thoughts to aid grooming.

1) what's the motivation?

A) datasets large enough to fit into autogluon for subsampling, but too big for the models to train on?
B) datasets so big they won't even fit into memory for the AG subsampling stage, so subsampling needs to be batched or piped
C) fits in memory in models but we think we can train faster on undersampled data, possibly even giving us a better evaluation_metric in the time budget

2) How do we want this implemented? Multiple options may be supported :

A) Naive random undersampling - quick and low complexity means low chance of blowing up memory during undersampling
B) clever enough to reduce class imbalances, with random removal of majority classes
C) not only improving class imbalance, but intelligent choice of which samples to remove (prototype selection) OR construction of new averaged samples (prototype generation, see https://imbalanced-learn.org/stable/under_sampling.html).  May itself blow up memory or take material time for the undersampling. 
D) Some other option I'm not aware of. 


Nicest would be a generic module that does 2A+B+C with functionality chosen by caller. 

E.g. If we detect extreme memory pressure, maybe caller asks for 2A or 2B with a moderate reduction ratio, then maybe 2C to reduce further. 

If we want something out soon let's do 2A or 2A+B while possibly defining an API that supports 2C snd returns "not implemented". 


3 Do we support class imbalance reduction for 

A) only classification
B) also regression.  Does the user input the expected distribution function?

**Comment by Innixma** [comment]
@willsmithorg 

> what's the motivation?
A) datasets large enough to fit into autogluon for subsampling, but too big for the models to train on?
B) datasets so big they won't even fit into memory for the AG subsampling stage, so subsampling needs to be batched or piped
C) fits in memory in models but we think we can train faster on undersampled data, possibly even giving us a better evaluation_metric in the time budget

I believe A will be the initial target, while B is a relatively easy add-on to A.

C is very interesting and is the backbone idea behind FLAML, but its a **big** undertaking. I'd guess without being very clever any naive implementation will simply be worse the vast majority of the time than not undersampling at all.


> How do we want this implemented? Multiple options may be supported :
A) Naive random undersampling - quick and low complexity means low chance of blowing up memory during undersampling
B) clever enough to reduce class imbalances, with random removal of majority classes
C) not only improving class imbalance, but intelligent choice of which samples to remove (prototype selection) OR construction of new averaged samples (prototype generation, see https://imbalanced-learn.org/stable/under_sampling.html). May itself blow up memory or take material time for the undersampling.
D) Some other option I'm not aware of.

A should be the way to go to solve the simple problem of getting the data to be small enough to train models with. We can then benchmark any smarter methods and compare against A in terms of final performance. B is an easy 'smarter' method to try, but it also depends on the metric whether we care about class imbalance (for example accuracy could get worse if we fix imbalance).

C is the big one, there are many papers on this and ideally if there are drop-in techniques we can give them a try and compare, but I'd only do this after A & B are investigated.

> 3 Do we support class imbalance reduction for
A) only classification
B) also regression. Does the user input the expected distribution function?

Probably just classification initially, I doubt many users would know the distribution if it differed from their training data, but it is something to consider in future.

For first steps, I'd say 2A to start, with 2B as a fast-follow (need to better understand what format someone with huge data is loading from, parquet? or maybe CSV if they are crazy)
